### DISCUSSION TITLE
📢关于课程《第一课：创建第一个DApp》的讨论

### DISCUSSION THREAD

**Discussion by tianxiuali** [Ed3Academy/blockchain-courses#13] [课程讨论]
欢迎在这里，对《第一课：创建第一个DApp》这门课程，提出你的疑问或建议😄

**Comment by Gereey** [comment]
“升级你的门户网站”小节
![截图](https://user-images.githubusercontent.com/104134709/225618056-d0652b46-0796-4d84-bb2e-6aca25de6eee.jpg)
是要在哪里先定义吗？

**Reply by tianxiuali** [reply]
感谢同学的提问！
这小节课程确实遗漏了对mintTo的定义和使用😅
现在已经补充完整，请再试试运行你的项目。
![image](https://user-images.githubusercontent.com/17540361/225797230-09162111-3c13-4630-8f18-4f570c5fd8e2.png)

**Reply by Gereey** [reply]
## 感谢回复，还有两个疑问：

### 1. value那里显示类型不匹配，

![截图1](https://user-images.githubusercontent.com/104134709/225805520-95622222-8c6b-409d-b77e-224a989fb1af.jpg)
 #### 可以直接强制类型转换为Number吗？
### 2. 部署到测试网上时显示合约构造函数参数缺失
![image](https://user-images.githubusercontent.com/104134709/225805450-ec824c9b-ce4c-4b69-8e6e-a8adb9e2b607.png)
#### 合约构造函数
![image](https://user-images.githubusercontent.com/104134709/225807786-38103372-e429-4ae3-bf44-aa887f262f47.png)

#### 该如何解决呢？

**Reply by naruto3315** [reply]
HelloHello，同学好😀！问题解答如下
1. 需要通过指定返回值类型来解决，具体步骤如下：
    a. 安装 ethers 依赖
    ```cmd
    npm install ethers@5.7.2
    ```
    b. 在 `api\index.tsx` 中导入 ethers 依赖
    ```tsx
    import { BigNumber } from "ethers";
    ```
    c. 修改一下 `mintPrice` 的读取方式
    ```tsx
    const { data: mintPrice } = useContractRead<
      typeof abi,
      "mintPrice",
      number | BigNumber | undefined
    >({
      address: contractAddress,
      abi,
      functionName: "mintPrice",
    });
    ```
2. 构造函数新增了一个参数 `mintPrice`，所以要**修改部署脚本调用构造函数的地方**，如下
```
  let mintTokenNumber = 10;
  let mintPrice = 1000;

  const myTokenContractFactory = await hre.ethers.getContractFactory("MyToken");
  const myTokenContract = await myTokenContractFactory.deploy(
    mintTokenNumber,
    mintPrice
  );
  await myTokenContract.deployed();
```

**Reply by Gereey** [reply]
# 十分感谢解答，成功了！！！

![nft](https://user-images.githubusercontent.com/104134709/225883066-f156223c-1c23-4676-800e-6c8e613d983b.jpg)

**Reply by Gereey** [reply]
### 1. 请问我们应该在哪访问我们的门户网站呢？在polygonscan里面通过搜索合约地址看到的就是么？该如何邀请别人领取token呢？
### 2. 还有就是NFT是发放到哪个网络上呢？metamask连接到polygonMainnet没有看到，是要手动添加吗？

**Reply by linyuliang** [reply]
1. 门户网站？指的是ed3academy.xyz么？目前通过谷歌搜索和直接官网地址进入。NFT合约是发布再Polygon的主网，地址是：https://polygonscan.com/token/0x5d1cb4b03a2fde2671987563ffeb8234d40ae8eb。 目前如果是要别人也能领取到NFT，需要别人进入ed3academy.xyz完成课程，例如通过海报图片引入我们官网；如果是门户网站指的是这门课程的前端网站，你可以选择使用免费的vercel来发布前端站点，然后别人通过你的前端站点vercel域名，访问你的dapp服务。但是目前vercel的域名是需要vpn才能访问，如果要国内可访问，需要申请域名DNS，A记录映射到vercel域名。
2.  我们课程凭证的NFT是在Polygon主网上的，可以在opensea上搜索NFT合约或者你自己的账户地址看到，metamask钱包的投资组合网站进去也可以看到，通常其他的NFT市场也都可以看到，因为是通用的ERC1155合约。
![2023 03 17_林玉良 fd03c66e2da43541b831c528c1e92fe3](https://user-images.githubusercontent.com/3059826/225942222-ce938ec4-0cc9-4168-bc30-97a92873a0d0.png)
![2023 03 17_林玉良 9f895b373a6f7979d88f2cf11447e633](https://user-images.githubusercontent.com/3059826/225942373-7c8eea92-d1bb-4e74-b35c-6796a3e1564d.png)

**Reply by Gereey** [reply]
# 好的，十分感谢回答。
